### PR TITLE
RFC6062: Implement client side for TCP allocations

### PR DESCRIPTION
**Pull request by AndyLc**
#### Description
Builds off of https://github.com/pion/turn/pull/148

Implements client side support for TCP allocations.
I'm also planning on adding support to server side

For the TCP allocation implementation, I tried follow the pattern of creating a client, calling allocate, and then using that allocation for communication. I thought that instead of calling WriteTo() and ReadFrom() on a TCP allocation, it would be better to Accept and Dial, and return a net.Conn for the end user to use, so I ended up having to do some refactoring.

Any feedback is welcome!

#### Usage:
1. Start coturn or some other turn server 
`./bin/turnserver --lt-cred-mech -u andy:secret -r pion.ly --no-cli ./examples/ca/turn_server_cert.pem --pkey ./examples/ca/turn_server_pkey.pem --allow-loopback-peers`

2. Run tcp client, also starting up a simple tcp server to communicate relay IPs.
`./tcp_alloc --host 127.0.0.1 --user andy=secret -server` 

3. Run another tcp client, which gets the relay IP from the other client and sends its own.
`./tcp_alloc --host 127.0.0.1 --user andy=secret`
 
Both clients should read and write a single message.

#### Reference issue
https://github.com/pion/turn/issues/143, https://github.com/pion/turn/issues/118

### CODE REVIEW

**Comment by rg0now** [comment]
Thanks for the PR, I'm loving it! I also wanted to take up #148 myself, I'm very happy that you came forward to do that.

Quick question: what is the purpose of the new `Protocol` field in the `ClientConfig`? My initial guess was that this was to fix the client exclusively to RFC6062-TCP allocations, but then why do you introduce an `AllocateTCP` function for a seemingly similar purpose? I mean, what's the difference between calling `Allocate` on a client with `client.Protocol="tcp"` and calling `AllocateTCP` on a client with `client.Protocol="udp"`? Then it occurred to me that most probably you introduce this in order to drive the client in resolving the TURN/STUN server addresses, but this is not needed: RFC6062-TCP allocations get you a connection to a TCP *peer* and not a TCP STUN/TURN *server*. 

My suggestion would be to remove `ClientConfig.Protocol`  and let the user specify the peer protocol by calling either `Allocate` or `AllocateTCP`. Also, you won't need a separate UDP/TCP resolving round for `stunServ` and `turnServ`, just use `ResolveUDPAddr`, this is what we have used so far anyway.

I'll try to do a formal review later too. And thanks again for picking this PR up, it is really useful!

**Comment by AndyLc** [comment]
> Thanks for the PR, I'm loving it! I also wanted to take up #148 myself, I'm very happy that you came forward to do that.
> 
> Quick question: what is the purpose of the new `Protocol` field in the `ClientConfig`? My initial guess was that this was to fix the client exclusively to RFC6062-TCP allocations, but then why do you introduce an `AllocateTCP` function for a seemingly similar purpose? I mean, what's the difference between calling `Allocate` on a client with `client.Protocol="tcp"` and calling `AllocateTCP` on a client with `client.Protocol="udp"`? Then it occurred to me that most probably you introduce this in order to drive the client in resolving the TURN/STUN server addresses, but this is not needed: RFC6062-TCP allocations get you a connection to a TCP _peer_ and not a TCP STUN/TURN _server_.
> 
> My suggestion would be to remove `ClientConfig.Protocol` and let the user specify the peer protocol by calling either `Allocate` or `AllocateTCP`. Also, you won't need a separate UDP/TCP resolving round for `stunServ` and `turnServ`, just use `ResolveUDPAddr`, this is what we have used so far anyway.
> 
> I'll try to do a formal review later too. And thanks again for picking this PR up, it is really useful!

Thank you for looking at the PR! I agree, specifying Protocol in the ClientConfig and having both AllocateTCP and AllocateUDP is redundant.

I'll update the PR with those changes.

**Comment by rg0now** [comment]
> it would be better to Accept and Dial, and return a net.Conn for the end user to use, so I ended up having to do some refactoring.

I'm thinking about the semantics of the thingie that is supposed to be returned from `AllocateTCP`:
- on the one hand it should be a proper `net.Dialer` so that the client can call `Dial` or `DialContext` to create a new connection to a peer
- on the other hand it also must implement the `net.Listener` interface so that we can call `Accept` on it to handle the cases when a peer connects to our transport relay connection on the server.

Wdyt?

**Comment by AndyLc** [comment]
I think that makes a lot of sense. What I return right now supports just Dial() and Accept()

Thinking about making it implement net.Listener, since `net.Listener`'s `Accept()` returns `(Conn, error)`, if I wanted to know the address of the peer that is trying to connect to me, calling RemoteAddr() will give me the address of the TURN server, not the peer. We would need to create a new struct which implements net.Conn that returns the correct RemoteAddr().

As for net.Dialer, what are you thinking for making it a net.Dialer but also implementing net.Listener?

**Comment by rg0now** [comment]
> Thinking about making it implement net.Listener, since net.Listener's Accept() returns (Conn, error), if I wanted to know the address of the peer that is trying to connect to me, calling RemoteAddr() will give me the address of the TURN server, not the peer. We would need to create a new struct which implements net.Conn that returns the correct RemoteAddr().

This is very similar to how [the `LocalAddr` function of the `UDPConn`](https://github.com/pion/turn/blob/7abfa3bf2628e2869945114ce06b8c2a92b0438e/internal/client/conn.go#L304) returned by `client.Allocate` actually returns the relay transport address instead of the local address. So indeed, this is a good idea.

> As for net.Dialer, what are you thinking for making it a net.Dialer but also implementing net.Listener?

The client needs to be able to both (1) accept new connections from different peers and (2) make new connections to different peers.

This is simple for UDP that is unconnected: we have a single `Allocate` returning a "universal" `UDPConn` that you can use to send to *any* peer by specifying the peer address in the `WriteTo`, and to receive packets from any peer by using `ReadFrom`. Problem is: this does not exist with TCP since there is no `WriteTo` and `ReadFrom`, only `Write` and `Read`, as TCP is connected. So every time the client wants to make a new connection to a peer it has to create a new TCP connection via the TURN server, which is essentially the same as implementing `net.Dialer` on the thing returned by `AllocateTCP`, and every time it wants to accept new connections from a peer it has to make a new connection again, which maps nicely to `net.Listener`.

So what I'm saying is that, for a client to accept new connections from a peer it should call `Accept` on the thing returned by `AllocateTCP`:

```
stuff, err := client.AllocateTCP()

conn, err := stuff.Accept()
```

While for making a new connection to a peer it has to call `Dial`:

```
stuff, err := client.AllocateTCP()

conn, err := stuff.Dial("tcp", peerAddr)
```

I'm thinking about a good name for `stuff`: `transport` or `net` are the closest I can come up with right now, as per `pion/transport`...

**Comment by AndyLc** [comment]
Makes sense, I totally agree

> I'm thinking about a good name for `stuff`: `transport` or `net` are the closest I can come up with right now, as per `pion/transport`...

I'm open to either, maybe `socket`? It feels like we're using a socket on the allocation we made to accept to connections from peers, or dial to peers.

**Comment by AndyLc** [comment]
> * create a new `turn.Dialer` interface that generalizes `net.Dialer` (I can't for the love of god understand why `net.Dialer` is not an interface but anyway), which is essentially a factory to make new connections (whatever type of connections, doesn't have to be TCP):
>   ```
>   type Dialer interface {
>       Dial(network, address string) (Conn, error)
>       DialContext(ctx context.Context, network, address string) (Conn, error) // I'm not sure we need this...
>   }
>   ```
> * make a default implementation that implements `turn.Dialer` using `net`
> * allow the user to pass in their own implementation either in the `NewClient` or the `AllocateTCP` call (I'd prefer the first) but make this optional; if no `turn.Dialer` is specified then we set the default implementation that uses `net`
> * convert `TCPConn` to use the abstract `Dialer`

Sounds great, I will add this. One issue, currently the turn package imports client. `TCPConn` lives in package client, and if I were to give it a user-defined dialer, it would create create circular dependency. Shall we move turn.Dialer to a new package? Maybe utils or something?

Also for TCPRelay, you're thinking renaming `TCPConn`->`TCPRelay`? Or creating another interface, TCPRelay, which has both Dial and Accept?

**Comment by rg0now** [comment]
> Sounds great, I will add this. One issue, currently the turn package imports client. TCPConn lives in package client, and if I were to give it a user-defined dialer, it would create create circular dependency. Shall we move turn.Dialer to a new package? Maybe utils or something?

Hmm, good catch. My gut feeling is that it should go into `pion/transport` but I really  don't know. Let's put it into `internals/ipnet` for now but this is not optimal, I hope @stv0g will jump in and suggest something better.

> Also for TCPRelay, you're thinking renaming TCPConn->TCPRelay? Or creating another interface, TCPRelay, which has both Dial and Accept?

I'd call the new interface that is both a `net.Dialer` and a `net.Listener` as `TCPRelay`, I would put this into `internals/ipnet` or a new package `internals/net` or `internals/util` or `internals/transport`, and I would make `TCPConn` an implementation of this interface. Does this sound stupid? API-design-wise this PR ended up pretty unique: I've never seen a case when a TCP endpoint behaved both as a `net.Dialer` and a `net.Listener`.

**Comment by rg0now** [comment]
As per why `net.Dialer` is not an interface, see this: https://github.com/golang/go/issues/9360. I don't agree with the points made there, but anyway.

**Comment by AndyLc** [comment]
> let's put the `Relay` interface into a new package called `internal/transport` and then the users of it would refer to it as `transport.Relay` and that would nicely connote the meaning that this interface works in two directions, both as a Dialer and a Listener.

So I tried creating this interface
```
import (
	t "github.com/pion/transport/v2"
	"net"
)

type Relay interface {
	net.Listener
	t.Dialer
}
```
but net.Listener's Accept() only returns a Conn and error. From net.Conn, RemoteAddr() will return the address of the turn server, not the peer. Do you have any ideas on the best way to convey from which peer we are accepting a connection?

Originally I had Accept also return another address, but maybe there's a better way?

**Comment by rg0now** [comment]
> but net.Listener's Accept() only returns a Conn and error. From net.Conn, RemoteAddr() will return the address of the turn server, not the peer. Do you have any ideas on the best way to convey from which peer we are accepting a connection?

Bad news: you'll have to implement your own `net.TCPListener` and `net.Conn` for this. Good news, this seems simple thanks to Go's embedding. So here is my take.

Basically we need to create our own implementation of `net.Listener` called, say, `TCPRelay`, that will provide us our own `net.Conn` implementation, say, `TCPRelayConn`, on accept. 

```
type TCPRelay struct {
    net.TCPListener
    // put something here that will provide the custom remote addr for the connections created
    // on this TCPRelay, say, a turn.Allocation
}

type TCPRelayConn struct {
    net.Conn
    RemoteAddr net.Addr
}

// Implement all of net.Listener for TCPRelay, here is the Accept() for a sample
func (l *TCPRelay) Accept() (TCPRelayConn, error){
    c, err := TCPListener.Accept()
    a := <obtain the remote addr for this allocation>
    return &TCPRelayConn{c, a}, err
}

// Implement all of net.Conn for TCPRelayConn, here is RemoteAddr
func (c *TCPRelayConn) RemoteAddr() net.Addr {
    return c.RemoteAddr
}

// Implement the whole of `transport.Dialer` for TCPRelay
func (l *TCPRelay) Dial(network, address string) (TCPRelayConn, error) {
    // use DialTCP or similar and then return the TCPRelayConn that wraps the net.Conn obtained
}
```

At the end we should have a default implementation for the `Relay` interface that people can use if
they don't want to mess with overriding anything, but since your code internally only ever
interacts with the `Relay` interface it is generic enough to support people providing their own
implementation.

Looking at this right now, I'm not sure we have to take all this complexity at the first step. We
could go with your the hardcoded-TCP implementation for now, let the PRs get merged, and only after
that generalize? What do you think is the best way forward?

**Comment by AndyLc** [comment]
> We could go with your the hardcoded-TCP implementation for now, let the PRs get merged, and only after
> that generalize? What do you think is the best way forward?

Agreed, I think getting the PRs in and then generalizing will make the code changes simpler and more iterative. 

I'll re-push my current changes based off your suggestions now!

**Comment by stv0g** [comment]
Hi all, I am joining the party a bit late. So I will need some more time to review the PR. Thanks @AndyLc for starting to work on it :) And @rg0now for the first review :)

> [...] you'll have to implement your own net.TCPListener

We already have an interface for this defined here: https://github.com/pion/transport/blob/22a2f3b9fe3381cb4a2735420e05b64813ca3e4f/net.go#L378

Did you maybe already have a look on how we can plug this into pion/ice?
I think its important to keep this in mind before we merge an API which would be hard/impossible to be used by pion/ice

**Comment by rg0now** [comment]
You mean [`Agent.gatherCandidatesRelay`](https://github.com/pion/ice/blob/cf6f758882661476a2d9b7d6eb0a19d9ff74cab8/gather.go#L506), no?

The definite answer is in RFC 6544 ([TCP Candidates with Interactive Connectivity Establishment (ICE)](https://datatracker.ietf.org/doc/rfc6544) but I don't have to read it. I think the main idea is that we create *two* allocations on any TCP TURN server: one conventional UDP and another RFC 6062 style TCP. Then, we send two relay candidates over to the peer, one with protocol UDP and another with TCP. 

If we are at RFCs: here is what RFC 8835 (Transports for WebRTC) has to say on this:

TURN TCP candidates, where the connection from the WebRTC endpoint's TURN server to the peer is a TCP connection, [RFC6062] MAY be supported.

However, such candidates are not seen as providing any significant benefit, for the following reasons.

First, use of TURN TCP candidates would only be relevant in cases where both peers are required to use TCP to establish a connection.

Second, that use case is supported in a different way by both sides establishing UDP relay candidates using TURN over TCP to connect to their respective relay servers.

Third, using TCP between the WebRTC endpoint's TURN server and the peer may result in more performance problems than using UDP, e.g., due to head of line blocking.

**Comment by AndyLc** [comment]
Thanks @stv0g for taking a look!

> I think the main idea is that we create _two_ allocations on any TCP TURN server: one conventional UDP and another RFC 6062 style TCP. Then, we send two relay candidates over to the peer, one with protocol UDP and another with TCP.

If we go with this, I believe it should be pretty straightforward. In addition to calling Allocate(), we also call `AllocateTCP()` when `url.Proto == ProtoTypeTCP`. Looks like network type, which is always UDP needs to be changed too. I can take a more detailed look later.

One concern I had with this method is, I feel we should only gather TCP relay candidates under certain conditions. For example, if we ever created an offer specifying `m=audio 45664 TCP/RTP/AVP 0`, we should only generate TCP candidates right? In that case, we might need to rework `GatherCandidates()` to take in the target protocol, otherwise allow all.

**Comment by stv0g** [comment]
>  I feel we should only gather TCP relay candidates under certain conditions.

The ICE agent config already allows you to specify which types of candidates should be gathered via its `NetworkTypes` and `CandidateTypes` settings.

**Comment by rg0now** [comment]
I took a look at the client API proposed in #143. Turns out there are some important differences.

1. In order for a client to obtain a new TCP connection to a peer, in this proposal the client uses the "thing" returned from `AllocateTCP` as a Dialer:

   ```
   relayConn, _ := client.AllocateTCP()
   conn, _ = relayConn.Dial("tcp", peerAddr)
   ```

   In contrast, in #143 we first make a TCP allocation on the server, obtain a new connection id, we open another TCP connection, and finally bind the new connection to the connection id we have just obtained in a separate step:

   ```
   // not sure what to do with relayConn
   relayConn, _ := client.AllocateTCP()
   
   // attempt connection, get connection id
   cid, _ := client.Connect(peerAddr)
   
   // open new data connection to turn server
   dconn, _ := net.Dial("tcp", turnServerAddr)
   
   // associate connection with connection id
   _ = client.ConnectionBind(dconn, cid)
   ```

   I personally prefer the latter, since it allows to bind *any* `net.Conn` to the connection id, not just `net.TCPConn`. This would be fairly difficult with the API proposed here, where the `net.(TCP)Conn` is created and returned by the Dialer (see the discussion above).

2. To support the opposite case, when it is the peer that initiates the new connection, the client has to be able to call `Accept` on the thing returned by `AllocateTCP`. I don't see support for this in #143 (observe the comment "not sure what to do with relayConn" in the above snipppet, I took it literally from the PR). This PR nicely handles this by returning a thing that is also a proper `net.Listener`.

In summary, I prefer the API of #143 for making client->peer connections since it does not require us to make the thing returned from `AllocateTCP` a Dialer, and the API proposed here for making peer->client connections as it is a proper `net.Listener`. Wdyt?

@AndyLc: if we adopt this new client API then my advice to implement `t.Relay` is void. I hope it was at least a fun experience, sorry for that...:-(

**Comment by AndyLc** [comment]
> 1. I personally prefer the latter, since it allows to bind _any_ `net.Conn` to the connection id, not just `net.TCPConn`. This would be fairly difficult with the API proposed here, where the `net.(TCP)Conn` is created and returned by the Dialer (see the discussion above).

@rg0now Hm, is the benefit you're getting at that it would be easier to create your own implementation of net.Conn rather than transport.Dialer?

I'm also utilizing a user chosen transport.Dialer in `HandleConnectionAttempt`, where I need to create a new connection to the relay to return when the user calls Accept(). So, if I were to remove Dial from the `t.Relay`, the user would still need an implementation of Dialer if they want custom functionality.

**Review by stv0g — CHANGES_REQUESTED** [review]
Hi @AndyLc 

Thanks a lot for your work. It looks great so far.
I mainly have a few nitpick comments.

This is a first round of review. I will probably have a few more comments later.

One thing, we should not forget is to add some info the README.md as well.

**Comment by stv0g** [comment]
Hi @rg0now,

> Also, can you please comment on the API? It'd be nice to get your opinion on https://github.com/pion/turn/pull/311#issuecomment-1498162747.

Sure. I had to spend some time to think about it.
Currently, I am leaning towards the API proposed by @AndyLc in this PR in favor of #143.
Mainly, because I find it hard to find a use case for binding an existing TCP connection to the connection ID.
It just makes the API harder to use as the user must establish the TCP data connection to the same TURN server address as the control connection.

But I am not against supporting also the ConnectionBind use case.

Here is my proposal which basically combines the two approaches by moving the `client. ConnectionBind()` method to `allocation.BindConnection()`:

```go
package turn

type Client struct { ... }

func (c *Client) AllocateTCP() (*TCPAllocation, error)

type TCPAllocation struct {
	// ConnectionID uniquely identifies a peer data connection.
	ConnectionID uint32

	// addr is the relayed transport address
	addr *net.TCPAddr
	client *Client

	...
}

// BindConnection associates the provided connection as the allocations data connection.
func (a *TCPAllocation) BindConnection(conn net.Conn) error

// Dial connects to the address on the named network.
func (a *TCPAllocation) Dial(network, address string) (net.Conn, error) {}

// DialContext connects to the address on the named network using the provided context.
func (a *TCPAllocation) DialContext(ctx context.Context, network, address string) (net.Conn, error)

func (a *TCPAllocation) DialTCP(network string, laddr, raddr *net.TCPAddr) (*TCPConn, error) {
	...
	a.ConnectionID = ...

	conn, err := net.DialTCP(network, client.BindConnection)
	if err != nil {
		return nil, fmt.Errorf("failed to establish connection to TURN server: %w", err)
	}

	if err := a.BindConnection(conn); err != nil {
		return nil, fmt.Errorf("failed to bind connection: %w", err)
	}

	return &TCPConn{
		TCPConn: conn,
		allocation: a,
	}, nil
}

// Accept waits for and returns the next connection to the listener.
func (a *TCPAllocation) Accept() (net.Conn, error) { ... }

// AcceptTCP accepts the next incoming call and returns the new connection.
func (a *TCPAllocation) AcceptTCP() (*TCPConn, error) { ... }

// Addr returns the allocations relayed address.
func (a *TCPAllocation) Addr() net.Addr { ... }

// Close releases the allocation
// Any blocked Accept operations will be unblocked and return errors.
// Any opened connection via Dial/Accept will be closed.
func (a *TCPAllocation) Close() error { ... }

// SetDeadline sets the deadline associated with the listener.
// A zero time value disables the deadline.
func (a *TCPAllocation) SetDeadline(t time.Time) error { ... }

// TCPConn wraps a net.TCPConn and returns the allocations relayed
// transport address in response to TCPConn.LocalAddress()
type TCPConn struct {
	*net.TCPConn

	allocation *TCPAllocation
}

func (c *TCPConn) LocalAddress() net.Addr {
	return c.allocation.Addr()
}

var _ transport.TCPListener = (*TCPAllocation)(nil) // Includes type check net.Listener
var _ transport.TCPConn = (*TCPConn)(nil) // Includes type check for net.Conn
```

**Comment by stv0g** [comment]
I currently see no need to define our own interface types here as I can not see anything which would need to be user customizable.

In my previous post I've choosen some new naming:

- `RelayConn` -> `TCPAllocation` 

Also, @AndyLc let me know if I can support you with the implementation :)

**Comment by rg0now** [comment]
I really like this API: clean and simple.

```go
var _ transport.TCPListener = (*TCPAllocation)(nil) // Includes type check for net.Listener
var _ transport.TCPConn = (*TCPConn)(nil) // Includes type check for net.Conn
var _ transport.Dialer = (*TCPAllocation)(nil)
```

Huhh, it didn't occur to me that you could do this in Go, nice! I guess we want to put this into the tests, no?

@AndyLc is there anything I can help you to get this PR in a mergeable form? I'd really love to have this feature available in pion/turn. Also, sorry for making you run these unnecessary circles with `net.Relay`, I hope it was not completely useless

File: client_test.go [hunk 1]
```diff
@@ -187,3 +187,53 @@ func TestClientNonceExpiration(t *testing.T) {
...
+		Password:       "pass",
+	})
+	assert.NoError(t, err)
+	assert.NoError(t, client.Listen())
+
+	allocation, err := client.AllocateTCP()
+	assert.NoError(t, err)
+
+	// TODO: Implement server side handling of Connect and ConnectionBind
+	// _, err = allocation.Dial(&net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 8080})
```
[the remark this answers is not in the document]
Comment (AndyLc) [reply client_test.go:232]: We are supporting both allocation.Dial() and allocation.Listen(). I wanted to make this test actually interact with the server-side implementation, but that will be in a separate PR, that is why I put the TODO there

File: internal/server/turn.go [hunk 2]
```diff
@@ -53,14 +53,14 @@ func handleAllocateRequest(r Request, m *stun.Message) error {
...
 	//    or is malformed, the server rejects the request with a 400 (Bad
 	//    Request) error.  Otherwise, if the attribute is included but
-	//    specifies a protocol other that UDP, the server rejects the
+	//    specifies a protocol other that UDP/TCP, the server rejects the
 	//    request with a 442 (Unsupported Transport Protocol) error.
 	var requestedTransport proto.RequestedTransport
 	if err = requestedTransport.GetFrom(m); err != nil {
 		return buildAndSendErr(r.Conn, r.SrcAddr, err, badRequestMsg...)
-	} else if requestedTransport.Protocol != proto.ProtoUDP {
+	} else if requestedTransport.Protocol != proto.ProtoUDP && requestedTransport.Protocol != proto.ProtoTCP {
```
[the remark this answers is not in the document]
Comment (AndyLc) [reply internal/server/turn.go:61]: The reason I added this was so in client_test.go, AllocateTCP() would not return an error. 

```
allocation, err := client.AllocateTCP()
require.NoError(t, err)
```

Maybe we can remove this error check and add it back in the server-side PR?

File: internal/client/tcp_conn.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+			relayedAddr: config.RelayedAddr,
+			permMap:     newPermissionMap(),
+			integrity:   config.Integrity,
+			_nonce:      config.Nonce,
+			_lifetime:   config.Lifetime,
+			log:         config.Log,
+		},
+	}
+
+	c.log.Debugf("initial lifetime: %d seconds", int(c.lifetime().Seconds()))
```
[the remark this answers is not in the document]
Comment (AndyLc) [reply internal/client/tcp_conn.go:52]: Hm in this repo there are many instances of c.log.Debug that use lower case, and I was following that. Should we change the standard to first letter uppercase?

File: internal/client/conn.go [hunk 4]
```diff
@@ -45,49 +46,60 @@ type UDPConnObserver interface {
...
 
-// UDPConn is the implementation of the Conn and PacketConn interfaces for UDP network connections.
-// comatible with net.PacketConn and net.Conn
-type UDPConn struct {
-	obs               UDPConnObserver       // read-only
+type RelayConn interface {
+	CreatePermissions(addrs ...net.Addr) error
+}
+
+type RelayConnContext struct {
```
[the remark this answers is not in the document]
Comment (AndyLc) [reply internal/client/conn.go:64]: RelayX are used for both TCP and UDP connections. Both UDPConn and TCPConn use the fields in RelayConnContext. In this case, do you prefer the naming of RelayConnImpl?

In client.go, RelayConn was used to generalize between UDPConn and TCPConn. I think we can get rid of this and each client will have a UDPConn and TCPAllocation.

**Comment by AndyLc** [comment]
Hey guys, I made some modifications to move towards @stv0g's proposal. There were some differences like 1. I think BindConnection needs to take in a ConnectionID, the reason being that under the same allocation, you can have multiple connections and thus multiple data connections w/ different connection IDs. So when you call BindConnection, you will need to specify which connection to bind.

Also, the CreatePermission in client.go may need to be reworked to specify which allocation to create a permission on, TCP or UDP. As it is now, it tries to create permissions on both. What would you guys prefer?

I also was unclear where to add these
``
`var _ transport.TCPListener = (*TCPAllocation)(nil) // Includes type check for net.Listener
var _ transport.TCPConn = (*TCPConn)(nil) // Includes type check for net.Conn
var _ transport.Dialer = (*TCPAllocation)(nil)
```
I assumed it was a type check so I just put it at the top of tcp_conn.go

**Comment by stv0g** [comment]
> I think `BindConnection` needs to take in a `ConnectionID`, the reason being that under the same allocation, you can have multiple connections and thus multiple data connections w/ different connection IDs. So when you call `BindConnection`, you will need to specify which connection to bind.

Oh, okay that is a detail which I missed. But it makes total sense.. Everytime, we `Accept` we basically get a new connection id. I've put some more thought into the API and updated my previous post with an updated API proposal.

The main change is the removal of `TCPAllocation.BindConnection()`. I've replaced it with additional methods:

- `TCPAllocation.DialTCPWithConn(conn net.Conn, network string, laddr, raddr *net.TCPAddr) (*TCPConn, error)`
- `TCPAllocation.AcceptTCPWithConn(conn net.Conn) (*TCPConn, error)`

Furthermore, the `ConnectionID` member moved from the `TCPAllocation` to `TCPConn` type.

> I also was unclear where to add these [...]

I would add those to the same place where the types which are checked are declared.

**Comment by stv0g** [comment]
The beauty of this API is, that basically `TCPAllocation.DialTCP()` is just thin wrapper around `TCPAllocation.DialTCPWithConn()`.

**Comment by AndyLc** [comment]
Ah, that makes a lot of sense, I like it. Do you think it's necessary at the moment to have DialTCP or just Dial? If we want DialTCP it seems we'll have to have a TCPDialer in pion/transport

**Comment by stv0g** [comment]
> Do you think it's necessary at the moment to have DialTCP or just Dial?

I think we could start with `Dial` first and add `DialTCP` lateron if needed.

**Comment by AndyLc** [comment]
Cool, I added those changes, let me know what you think! There are some small differences

File: internal/client/conn.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (stv0g) [reply internal/client/conn.go:58]: I think the confusion stems from the fact that we do not have a `client.UDPAllocation` type.
If we would have it, `client.RelayConnContext` would be more appropriately named `client.Allocation`.

But in fact, the fields of `client.RelayConnContext`are basically just the clients per-allocation state.

[the remark this answers is not in the document]
Comment (stv0g) [reply internal/client/conn.go:58]: I am just confused why we have a per-allocation `_nonce` and and `integrity` fields here. Also the naming of `ConnObserver` is confusing.

I think this could need some refactoring as well.
- Rename `RelayConnContext` to `Allocation`
- Rename `ConnObserver` to `Client` (Just an interface within `internal/client` representing the public `Client` type in order to break cyclic dependencies)
- Move integrity and nonce fields from `RelayConnContext`  to the new `Client interface`
- Rename `conn.go` to `udp_conn.go`
- Move common code from `conn.go` to `allocation.go`

**Review by stv0g — CHANGES_REQUESTED** [review]

**Comment by stv0g** [comment]
Hi @AndyLc,

thanks for the fantastic work :-) I've reviewed your code and have some ideas for refactoring `RelayConnContext`.
I am sorry, that we are going through these rounds. This really hasnt been because of bad design decissions by you, but rather previous contributors.. We must refactor some of the naming the the internal client package.

**Comment by stv0g** [comment]
Please let me know if I can help you with it :) I would be happy to also push changes to your branch.

File: internal/client/permission.go [hunk 5]
```diff
@@ -32,53 +33,44 @@ type permissionMap struct {
 	mutex   sync.RWMutex
 }
 
+func addr2IPFingerprint(addr net.Addr) string {
```
[the remark this answers is not in the document]
Comment (AndyLc) [reply internal/client/permission.go:36]: I actually got this code from the server side, if you look in `internal/allocation/allocation.go`. It's used in the server code right now, maybe we need a separate helper?

**Comment by AndyLc** [comment]
@stv0g Yes feel free to make any changes you want! I just added you to the repo.

File: internal/client/permission.go [hunk 6]
```diff
@@ -32,53 +33,44 @@ type permissionMap struct {
...
+	switch a := addr.(type) {
+	case *net.UDPAddr:
+		return a.IP.String()
+	case *net.TCPAddr:
+		return a.IP.String()
+	}
+	return "" // should never happen
```
[the remark this answers is not in the document]
Comment (stv0g) [reply internal/client/permission.go:43]: Ah, `panic()` is forbidden in our library code.. Okay, but that case should anyway never happen...

File: internal/client/tcp_conn.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+var (
+	errInvalidTURNFrame    = errors.New("data is not a valid TURN frame, no STUN or ChannelData found")
+	errIncompleteTURNFrame = errors.New("data contains incomplete STUN or TURN frame")
+)
+
+const (
+	stunHeaderSize = 20
+)
+
+type TCPAllocation struct {
```
[the remark this answers is not in the document]
Comment (stv0g) [reply internal/client/tcp_conn.go]: Done via 17fc9f1f5c655ca2f2d7ec4247996eedc27f894e

File: internal/client/errors.go [hunk 8]
```diff
@@ -8,7 +8,8 @@ var (
 	errFake                                = errors.New("fake error")
 	errTryAgain                            = errors.New("try again")
 	errClosed                              = errors.New("use of closed network connection")
-	errUDPAddrCast                         = errors.New("addr is not a net.UDPAddr")
+	errTCPAddrCast                         = errors.New("addr is not a tcp address")
```
[the remark this answers is not in the document]
Comment (stv0g) [reply internal/client/errors.go]: Done via 3d71d371f94b07e74bbd578c0976c2925db0654b

**Comment by stv0g** [comment]
@AndyLc thanks for the repo access. I've addressed my remaining comments myself and also rebased your `dev` to the upstream `master`.

**Please reset your local branch as it gets messed up otherwise due to the rebase.**

Lets, see what the CI says..

**Comment by AndyLc** [comment]
@stv0g Thanks for making edits! I saw that some tests were still failing, so I pushed another commit to address those, mainly fixing golangci-lint issues and adding more unit tests.

**Comment by stv0g** [comment]
Hi @AndyLc :)

Great! I think we are getting close merging this into master 🥳 

I currently, see only one thing missing: We should add support for `pion/transport.Net` to the `Client` as use `net.Dial` for creating the TCP data-connection. Right now we are limited to the standard `net.Dial`. By using the interface from `pion/transport` we will gain the ability to test the feature with `pion/transport/vnet` as soon as it gains TCP support.

**Comment by AndyLc** [comment]
@stv0g Sounds good, I attempted to add vnet, let me know what you think!

**Comment by stv0g** [comment]
Hi @AndyLc,

cool :) I've found that we had two implementations of a client mock (`dummyConnObverver` and `dummyClient`). I refactored this into `mockClient`.

I also simplified this mock client by moving more of the static configuration attributes to `AllocationConfig`.
I've done so to keep the changes to the API surface of `turn.Client` to a minimum.

**Review by stv0g — APPROVED** [review]
Hi @AndyLc,

I approved the PR now. Please feel free to pull the trigger for merging it if you are okay with my latest changes.

**Review by stv0g — APPROVED** [review]
Hi @AndyLc,

I approved the PR now. Please feel free to pull the trigger for merging it if you are okay with my latest changes.

**Comment by AndyLc** [comment]
@stv0g Looks like I don't have access to merge the PR `Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.` Your changes look good to me though!
